core/org.openjdk.jmc.flightrecorder.rules.jdk/src/main/java/org/openjdk/jmc/flightrecorder/rules/jdk/memory/ZGCAllocationStallRule.java

```diff
@@ -0,0 +1,189 @@
+
+	public static final TypedResult<IQuantity> ZGC_ALLOCATION_STALL_TOTAL_DURATION = new TypedResult<>(
+			"zgcAllocationStallTotalDuration", //$NON-NLS-1$
+			JdkAggregators.ZGC_ALLOCATION_STALL_COUNT, UnitLookup.NUMBER, IQuantity.class);
+
+	public static final TypedResult<IQuantity> ZGC_ALLOCATION_STALL_PER_MINUTE = new TypedResult<>(
+			"zgcAllocationStallPerMinute", //$NON-NLS-1$
+			Messages.getString(Messages.ZGCAllocationStallRule_RATE),
+			Messages.getString(Messages.ZGCAllocationStallRule_RATE_LONG), UnitLookup.NUMBER, IQuantity.class);
+	public static final TypedPreference<IQuantity> ALLOCATION_STALL_INFO_LIMIT = new TypedPreference<>(
+			"allocation.stall.rate.info.limit", //$NON-NLS-1$
+			Messages.getString(Messages.ZGCAllocationStallRule_CONFIG_INFO_LIMIT),
+			Messages.getString(Messages.ZGCAllocationStallRule_CONFIG_INFO_LIMIT_LONG), NUMBER,
+			NUMBER_UNITY.quantity(10));
+
+	public static final TypedPreference<IQuantity> ALLOCATION_STALL_WARNING_LIMIT = new TypedPreference<>(
+			"allocation.stall.rate.warning.limit", //$NON-NLS-1$
+			Messages.getString(Messages.ZGCAllocationStallRule_CONFIG_WARN_LIMIT),
+			Messages.getString(Messages.ZGCAllocationStallRule_CONFIG_WARN_LIMIT_LONG), NUMBER,
+			NUMBER_UNITY.quantity(100));
+
+	private static final List<TypedPreference<?>> CONFIG_ATTRIBUTES = Arrays
+			.<TypedPreference<?>> asList(ALLOCATION_STALL_INFO_LIMIT, ALLOCATION_STALL_WARNING_LIMIT);
+
+	private static final Collection<TypedResult<?>> RESULT_ATTRIBUTES = Arrays.<TypedResult<?>> asList(
+			TypedResult.SCORE, ZGC_ALLOCATION_STALL_EVENTS, ZGC_ALLOCATION_STALL_LONGEST_DURATION,
+			ZGC_ALLOCATION_STALL_TOTAL_DURATION, ZGC_ALLOCATION_STALL_PER_MINUTE);
+
+	@Override
+	public String getId() {
+		return ZGC_RESULT_ID;
+	}
+
+	@Override
+	public String getTopic() {
+		return JfrRuleTopics.GARBAGE_COLLECTION;
+	}
+
+	@Override
+	public String getName() {
+		return Messages.getString(Messages.ZGCAllocationStall_RULE_NAME);
+	}
+
+	@Override
+	public Map<String, EventAvailability> getRequiredEvents() {
+		return REQUIRED_EVENTS;
+	}
+
+	private IResult getResult(
+		IItemCollection items, IPreferenceValueProvider valueProvider, IResultValueProvider resultProvider) {
+		long infoLimit = valueProvider.getPreferenceValue(ALLOCATION_STALL_INFO_LIMIT).clampedLongValueIn(NUMBER_UNITY);
+		long warningLimit = valueProvider.getPreferenceValue(ALLOCATION_STALL_WARNING_LIMIT)
+				.clampedLongValueIn(NUMBER_UNITY);
+
+		IQuantity zgcAllocationStallCount = items.getAggregate(JdkAggregators.ZGC_ALLOCATION_STALL_COUNT);
+		IQuantity zgcAllocationStallTotalDuration = items.getAggregate(JdkAggregators.TOTAL_ZGC_ALLOCATION_STALL);
+		IQuantity zgcAllocationStallLongestDuration = items.getAggregate(JdkAggregators.LONGEST_ZGC_ALLOCATION_STALL);
+		if (zgcAllocationStallCount != null && zgcAllocationStallCount.doubleValue() > 0) {
+
+			//Calculate time after JVM Start
+			IQuantity timeAfterJVMStart = RulesToolkit.getEarliestStartTime(items)
```

Comment:
Come to think of it - this will not work well for OldObjectSample events. They are a bit special and may be timestamped far earlier than the current recording period. I think we might want to add a RulesToolkit function that explicitly avoid the OldObjectSample events for this...